[BACK-3177] [BACK-3176] EHR Settings updates

toddkazakov opened the pull request:
- Allow specifying custom cadence for EHR scheduled reports
- Add configuration for patient tags associations when patients are created from an EHR integration
- Refactor redox order matching to allow for handling account creation and subscription in a single order

ewollesen reviewed (changes requested):
I might be really confused about the goal of all this, but here are my thoughts:

There are no tests or any input validation added. Can we be very confident that this won't be called with a value like "0" or "2w" or "604800s".

It appears that the input is expected to be in [ISO-8601 Duration](https://en.wikipedia.org/wiki/ISO_8601#Durations) format, but that's not documented anywhere, so I don't know how to validate that this is doing what's expected. 

On top of that, the ticket referenced, BACK-3177 is about disabling reports, not specifying custom cadences, so... all in all, I think I need a more context before I can understand what's supposed to be happening here?

toddkazakov commented:
@ewollesen the input validation is enforced at the API level according to the API spec. The allowed values are listed here[https://github.com/tidepool-org/clinic/pull/160/files#diff-41f8b0b50297a7227a61ed2ada54e8c8d28b761983d650b2b1b53fcca1b8a76cR4741-R4745](url).

`0d` disables all reports which is what the linked ticket requires. We do have requests from clinics to support custom cadences - it felt right to add this functionality as part of this ticket.

ewollesen commented:
Hopefully you understand that my comments are here to try to explain my thinking. I see no problems with the PR as written, *except* that I don't have all the necessary context to be able to detect problems in the first place. Does that make sense?

> @ewollesen the input validation is enforced at the API level according to the API spec. The allowed values are listed [here](https://github.com/tidepool-org/clinic/pull/160/files#diff-41f8b0b50297a7227a61ed2ada54e8c8d28b761983d650b2b1b53fcca1b8a76cR4741-R4745).

Could you show me where that's enforced? I'm unable to find it anywhere. I saw that it's documented there, but I haven't been able to find the code that enforces or validates that declaration. Did GitHub hide some other generated file from me maybe?

What I was able to find is: [api/mappers.go#L739](https://github.com/tidepool-org/clinic/blob/tk-scheduled-report-cadence/api/mappers.go#L739). It converts any old string to a ScheduledReportsCadence, but it performs no parsing or validation.

I asked myself, "What would the code do if an invalid value were received?" This is unclear to me. I haven't been able to find any code within clinic that consumes the values (maybe it's not written yet, or in another project?), so I can't judge how a consumer of these values would react to an invalid value. I feel like that's the kind of thing I should be double-checking during a review.

So I did more sleuthing. I tried to understand the behavior of the code in response to these values. I noted that the four values from the spec you wrote [are (via code-generation) defined](https://github.com/tidepool-org/clinic/blob/tk-scheduled-report-cadence/api/gen_types.go#L670-L676), but aside from "14d", used when the presented value is blank, the others are not used anywhere that I could find. So there's no hints there for me on how they're intended to be used. I had hoped they'd be used in other (non-code-generation) places, and thereby lead me to sections of the code that could shed more insight, but I haven't been able to find any.

> `0d` disables all reports which is what the linked ticket requires. 

Thank you for clarifying that. I feel like that fact could be documented somewhere in a code comment or unit test—would you agree? What I discovered was that the constant value `N0d` isn't used anywhere in clinic. I saw from the ticket that having a way to disable the reports is a requirement, but I don't have the context to make the assumption that 0d means that reports are disabled. It makes sense once you say it, but I could just as easily assume it means "send it immediately/the same day". A value of "disabled" would have been clearer here—especially since the other values aren't used (again, that I've been able to find). Is a same-day report something in the cards? If so, consider if a value other than "0d" makes sense as disabled.

> We do have requests from clinics to support custom cadences - it felt right to add this functionality as part of this ticket.

Sure, and I see (now that you've explained) how you chose to implement disabling, via the 0d value, but that wasn't at all clear to me without your explanation. I don't have nearly the context around the ticket that you do.

toddkazakov commented:
@ewollesen As discussed in slack, input validation is performed by a middleware according to the schema. 

Juggling multiple tickets which change the schema is quite difficult because of the code generation tool will introduce conflicts, so I added the configuration options required by [BACK-3176] to this PR.

I believe I addressed some of your feedback, please review again.

toddkazakov commented on clinics/clinics.go:
```diff
@@ -153,6 +154,11 @@ type MRNSettings struct {
 	Unique   bool `bson:"unique"`
 }
 
+type TagsSettings struct {
```
`EHRTagsSettings` in `types.go` is automatically generated. Persisting models generated from the API spec is a recipe for disaster, because the code generator often changes its output.

toddkazakov commented on client/types.go:
```diff
@@ -909,11 +909,23 @@ type EHRSettings struct {
 // EHRSettingsProvider defines model for EHRSettings.Provider.
 type EHRSettingsProvider string
 
+// EHRTagsSettings This configuration only applies to integrations using Redox Data Model
```
Auto generated from the API spec. I'll leave as it is.

toddkazakov commented on client/types.go:
```diff
@@ -909,11 +909,23 @@ type EHRSettings struct {
 // EHRSettingsProvider defines model for EHRSettings.Provider.
 type EHRSettingsProvider string
 
+// EHRTagsSettings This configuration only applies to integrations using Redox Data Model
```
See https://github.com/tidepool-org/clinic/pull/160#discussion_r1791640720

toddkazakov commented:
/deploy qa2

ewollesen commented on redox/redox.go:
```diff
@@ -306,116 +321,76 @@ func (h *Handler) RescheduleSubscriptionOrdersForPatient(ctx context.Context, pa
+				if patient == nil || patient.UserId == nil {
+					continue
+				}
+				if _, ok := unique[*patient.UserId]; ok {
```
```suggestion
				if _, found := unique[*patient.UserId]; found {
```
`ok` is used in a lot of examples like this, and it's fine, but I've always felt that tailoring the term more specifically to the situation can help the reader better understand the intention.

ewollesen commented on spec/clinic.v1.yaml:
```diff
@@ -4564,6 +4564,36 @@ components:
+              enum:
+                - ENABLE_REPORTS
+                - DISABLE_REPORTS
+              description: Optional action to be performed when a unique match has been found
```
This is such a strange concept to me. Are there docs explaining the need/intent behind this?

Why would someone want to make a change to a patient's records without having a known unique identifier for said patient? Isn't there a larger than necessary risk of applying a change to the wrong record?

Perhaps the situation is a clinic worker that has a note that patient John Doe DOB 1969-01-01 should have this order performed... but I guess they might not know if that is unique or not... Hmm... I suppose that makes sense... They don't have to first search to see if it is unique, they can just put the order through, and if it succeeds, it was unique, and if not, they (presumably) get notified...

Perhaps I've answered my own questions then, but if I'm mistaken, please let me know.

ewollesen commented on redox/redox.go:
```diff
@@ -306,116 +321,76 @@ func (h *Handler) RescheduleSubscriptionOrdersForPatient(ctx context.Context, pa
 		return nil, err
 	}
+	for _, filter := range filters {
+		// When matches are non-unique, the results are not useful to clients.
```
The consumers of this function are generally looking for unique matches. During the normal course of operations, having a handful (say < 10) matches is conceivable, but having 100 matches, given the criteria, would strongly indicate an error or misconfiguration of a filter, or some other problem.

As such, and to make this comment clearer, I would suggest:
1. removing this comment
2. adding a GoDoc to the findMatchingPatients function like the following:

```
// findMatchingPatients based on a MatchOrder.
//
// The number of matching patients is limited to 100. It is expected that consumers of this
// function are searching for unique patient matches. It is normal for a small number (< 10)
// of patients to match, but 100 matches strongly indicates an unexpected filter behavior.
//
// Placing a limit of 100 prevents a misconfigured filter from needlessly returning all of a
// clinic's patients.
```

ewollesen reviewed (approved):
Looks OK. No blockers. Just questions and suggestions.

ewollesen reviewed (approved):
LGTM
